sharedbuffers/mapped_struct.py

```diff
@@ -2412,24 +2626,46 @@ def __init__(self, *args, **kwargs):
     def _make_empty(self):
         return ()
 
-    @cython.locals(mult = cython.long, x = cython.long, y = cython.long, len_ = cython.Py_ssize_t, i = cython.Py_ssize_t)
+    @cython.locals(
+        mult=cython.long,
+        x=cython.long,
+        y=cython.long,
+        len_=cython.Py_ssize_t,
+        i=cython.Py_ssize_t,
+        acc=cython.size_t,
+        lane=cython.size_t,
+    )
     def __hash__(self):
         if self._hash == -1:
             if cython.compiled and is_cpython:
-                # From Python 2.7 source code
-                mult = 1000003
-                x = 0x345678
-                len_ = len(self)
-                for i in xrange(len_):
-                    len_ -= 1
-                    y = hash(self[i])
-                    x = (x ^ y) * mult;
-                    mult += 82520 + len_ + len_;
-
-                x += 97531
-                if x == -1:
-                    x = -2
-                self._hash = x
+                if python38:
+                    # From Python 3.8 source code
+                    len_ = len(self)
+                    acc = XXPRIME_5
+                    for i in range(len_):
+                        lane = hash(self[i])
+                        acc += lane * XXPRIME_2
+                        acc = (acc << 31) | (acc >> 33)
+                        acc *= XXPRIME_1
+                    acc += len_ ^ (XXPRIME_5 ^ 3527539)
+                    if acc == -1:
```

Comment:
acc is size_t, which is unsigned, but you're comparing to a signed type. To get well defined behavior, you have to cast the literal:

```suggestion
                    if acc == cython.cast(cython.size_t, -1):
```